      <!DOCTYPE html>
      <html>
        <head>
            <meta charset="utf-8" />
            <title>SeqApiPopFastqToVcf</title>
            <style>.markdown-preview:not([data-use-github-style]) { padding: 2em; font-size: 1.2em; color: rgb(56, 58, 66); background-color: rgb(250, 250, 250); overflow: auto; }
.markdown-preview:not([data-use-github-style]) > :first-child { margin-top: 0px; }
.markdown-preview:not([data-use-github-style]) h1, .markdown-preview:not([data-use-github-style]) h2, .markdown-preview:not([data-use-github-style]) h3, .markdown-preview:not([data-use-github-style]) h4, .markdown-preview:not([data-use-github-style]) h5, .markdown-preview:not([data-use-github-style]) h6 { line-height: 1.2; margin-top: 1.5em; margin-bottom: 0.5em; color: rgb(0, 0, 0); }
.markdown-preview:not([data-use-github-style]) h1 { font-size: 2.4em; font-weight: 300; }
.markdown-preview:not([data-use-github-style]) h2 { font-size: 1.8em; font-weight: 400; }
.markdown-preview:not([data-use-github-style]) h3 { font-size: 1.5em; font-weight: 500; }
.markdown-preview:not([data-use-github-style]) h4 { font-size: 1.2em; font-weight: 600; }
.markdown-preview:not([data-use-github-style]) h5 { font-size: 1.1em; font-weight: 600; }
.markdown-preview:not([data-use-github-style]) h6 { font-size: 1em; font-weight: 600; }
.markdown-preview:not([data-use-github-style]) strong { color: rgb(0, 0, 0); }
.markdown-preview:not([data-use-github-style]) del { color: rgb(94, 97, 110); }
.markdown-preview:not([data-use-github-style]) a, .markdown-preview:not([data-use-github-style]) a code { color: rgb(82, 111, 255); }
.markdown-preview:not([data-use-github-style]) img { max-width: 100%; }
.markdown-preview:not([data-use-github-style]) > p { margin-top: 0px; margin-bottom: 1.5em; }
.markdown-preview:not([data-use-github-style]) > ul, .markdown-preview:not([data-use-github-style]) > ol { margin-bottom: 1.5em; }
.markdown-preview:not([data-use-github-style]) blockquote { margin: 1.5em 0px; font-size: inherit; color: rgb(94, 97, 110); border-color: rgb(209, 209, 210); border-width: 4px; }
.markdown-preview:not([data-use-github-style]) hr { margin: 3em 0px; border-top: 2px dashed rgb(209, 209, 210); background: none; }
.markdown-preview:not([data-use-github-style]) table { margin: 1.5em 0px; }
.markdown-preview:not([data-use-github-style]) th { color: rgb(0, 0, 0); }
.markdown-preview:not([data-use-github-style]) th, .markdown-preview:not([data-use-github-style]) td { padding: 0.66em 1em; border: 1px solid rgb(209, 209, 210); }
.markdown-preview:not([data-use-github-style]) code { color: rgb(0, 0, 0); background-color: rgb(234, 234, 235); }
.markdown-preview:not([data-use-github-style]) pre.editor-colors { margin: 1.5em 0px; padding: 1em; font-size: 0.92em; border-radius: 3px; background-color: rgb(240, 240, 240); }
.markdown-preview:not([data-use-github-style]) kbd { color: rgb(0, 0, 0); border-width: 1px 1px 2px; border-style: solid; border-color: rgb(209, 209, 210) rgb(209, 209, 210) rgb(193, 193, 194); border-image: initial; background-color: rgb(234, 234, 235); }
.markdown-preview[data-use-github-style] { font-family: "Helvetica Neue", Helvetica, "Segoe UI", Arial, freesans, sans-serif; line-height: 1.6; overflow-wrap: break-word; padding: 30px; font-size: 16px; color: rgb(51, 51, 51); background-color: rgb(255, 255, 255); overflow: scroll; }
.markdown-preview[data-use-github-style] > :first-child { margin-top: 0px !important; }
.markdown-preview[data-use-github-style] > :last-child { margin-bottom: 0px !important; }
.markdown-preview[data-use-github-style] a:not([href]) { color: inherit; text-decoration: none; }
.markdown-preview[data-use-github-style] .absent { color: rgb(204, 0, 0); }
.markdown-preview[data-use-github-style] .anchor { position: absolute; top: 0px; left: 0px; display: block; padding-right: 6px; padding-left: 30px; margin-left: -30px; }
.markdown-preview[data-use-github-style] .anchor:focus { outline: none; }
.markdown-preview[data-use-github-style] h1, .markdown-preview[data-use-github-style] h2, .markdown-preview[data-use-github-style] h3, .markdown-preview[data-use-github-style] h4, .markdown-preview[data-use-github-style] h5, .markdown-preview[data-use-github-style] h6 { position: relative; margin-top: 1em; margin-bottom: 16px; font-weight: bold; line-height: 1.4; }
.markdown-preview[data-use-github-style] h1 .octicon-link, .markdown-preview[data-use-github-style] h2 .octicon-link, .markdown-preview[data-use-github-style] h3 .octicon-link, .markdown-preview[data-use-github-style] h4 .octicon-link, .markdown-preview[data-use-github-style] h5 .octicon-link, .markdown-preview[data-use-github-style] h6 .octicon-link { display: none; color: rgb(0, 0, 0); vertical-align: middle; }
.markdown-preview[data-use-github-style] h1:hover .anchor, .markdown-preview[data-use-github-style] h2:hover .anchor, .markdown-preview[data-use-github-style] h3:hover .anchor, .markdown-preview[data-use-github-style] h4:hover .anchor, .markdown-preview[data-use-github-style] h5:hover .anchor, .markdown-preview[data-use-github-style] h6:hover .anchor { padding-left: 8px; margin-left: -30px; text-decoration: none; }
.markdown-preview[data-use-github-style] h1:hover .anchor .octicon-link, .markdown-preview[data-use-github-style] h2:hover .anchor .octicon-link, .markdown-preview[data-use-github-style] h3:hover .anchor .octicon-link, .markdown-preview[data-use-github-style] h4:hover .anchor .octicon-link, .markdown-preview[data-use-github-style] h5:hover .anchor .octicon-link, .markdown-preview[data-use-github-style] h6:hover .anchor .octicon-link { display: inline-block; }
.markdown-preview[data-use-github-style] h1 tt, .markdown-preview[data-use-github-style] h2 tt, .markdown-preview[data-use-github-style] h3 tt, .markdown-preview[data-use-github-style] h4 tt, .markdown-preview[data-use-github-style] h5 tt, .markdown-preview[data-use-github-style] h6 tt, .markdown-preview[data-use-github-style] h1 code, .markdown-preview[data-use-github-style] h2 code, .markdown-preview[data-use-github-style] h3 code, .markdown-preview[data-use-github-style] h4 code, .markdown-preview[data-use-github-style] h5 code, .markdown-preview[data-use-github-style] h6 code { font-size: inherit; }
.markdown-preview[data-use-github-style] h1 { padding-bottom: 0.3em; font-size: 2.25em; line-height: 1.2; border-bottom: 1px solid rgb(238, 238, 238); }
.markdown-preview[data-use-github-style] h1 .anchor { line-height: 1; }
.markdown-preview[data-use-github-style] h2 { padding-bottom: 0.3em; font-size: 1.75em; line-height: 1.225; border-bottom: 1px solid rgb(238, 238, 238); }
.markdown-preview[data-use-github-style] h2 .anchor { line-height: 1; }
.markdown-preview[data-use-github-style] h3 { font-size: 1.5em; line-height: 1.43; }
.markdown-preview[data-use-github-style] h3 .anchor { line-height: 1.2; }
.markdown-preview[data-use-github-style] h4 { font-size: 1.25em; }
.markdown-preview[data-use-github-style] h4 .anchor { line-height: 1.2; }
.markdown-preview[data-use-github-style] h5 { font-size: 1em; }
.markdown-preview[data-use-github-style] h5 .anchor { line-height: 1.1; }
.markdown-preview[data-use-github-style] h6 { font-size: 1em; color: rgb(119, 119, 119); }
.markdown-preview[data-use-github-style] h6 .anchor { line-height: 1.1; }
.markdown-preview[data-use-github-style] p, .markdown-preview[data-use-github-style] blockquote, .markdown-preview[data-use-github-style] ul, .markdown-preview[data-use-github-style] ol, .markdown-preview[data-use-github-style] dl, .markdown-preview[data-use-github-style] table, .markdown-preview[data-use-github-style] pre { margin-top: 0px; margin-bottom: 16px; }
.markdown-preview[data-use-github-style] hr { height: 4px; padding: 0px; margin: 16px 0px; background-color: rgb(231, 231, 231); border: 0px none; }
.markdown-preview[data-use-github-style] ul, .markdown-preview[data-use-github-style] ol { padding-left: 2em; }
.markdown-preview[data-use-github-style] ul.no-list, .markdown-preview[data-use-github-style] ol.no-list { padding: 0px; list-style-type: none; }
.markdown-preview[data-use-github-style] ul ul, .markdown-preview[data-use-github-style] ul ol, .markdown-preview[data-use-github-style] ol ol, .markdown-preview[data-use-github-style] ol ul { margin-top: 0px; margin-bottom: 0px; }
.markdown-preview[data-use-github-style] li > p { margin-top: 16px; }
.markdown-preview[data-use-github-style] dl { padding: 0px; }
.markdown-preview[data-use-github-style] dl dt { padding: 0px; margin-top: 16px; font-size: 1em; font-style: italic; font-weight: bold; }
.markdown-preview[data-use-github-style] dl dd { padding: 0px 16px; margin-bottom: 16px; }
.markdown-preview[data-use-github-style] blockquote { padding: 0px 15px; color: rgb(119, 119, 119); border-left: 4px solid rgb(221, 221, 221); }
.markdown-preview[data-use-github-style] blockquote > :first-child { margin-top: 0px; }
.markdown-preview[data-use-github-style] blockquote > :last-child { margin-bottom: 0px; }
.markdown-preview[data-use-github-style] table { display: block; width: 100%; overflow: auto; word-break: keep-all; }
.markdown-preview[data-use-github-style] table th { font-weight: bold; }
.markdown-preview[data-use-github-style] table th, .markdown-preview[data-use-github-style] table td { padding: 6px 13px; border: 1px solid rgb(221, 221, 221); }
.markdown-preview[data-use-github-style] table tr { background-color: rgb(255, 255, 255); border-top: 1px solid rgb(204, 204, 204); }
.markdown-preview[data-use-github-style] table tr:nth-child(2n) { background-color: rgb(248, 248, 248); }
.markdown-preview[data-use-github-style] img { max-width: 100%; box-sizing: border-box; }
.markdown-preview[data-use-github-style] .emoji { max-width: none; }
.markdown-preview[data-use-github-style] span.frame { display: block; overflow: hidden; }
.markdown-preview[data-use-github-style] span.frame > span { display: block; float: left; width: auto; padding: 7px; margin: 13px 0px 0px; overflow: hidden; border: 1px solid rgb(221, 221, 221); }
.markdown-preview[data-use-github-style] span.frame span img { display: block; float: left; }
.markdown-preview[data-use-github-style] span.frame span span { display: block; padding: 5px 0px 0px; clear: both; color: rgb(51, 51, 51); }
.markdown-preview[data-use-github-style] span.align-center { display: block; overflow: hidden; clear: both; }
.markdown-preview[data-use-github-style] span.align-center > span { display: block; margin: 13px auto 0px; overflow: hidden; text-align: center; }
.markdown-preview[data-use-github-style] span.align-center span img { margin: 0px auto; text-align: center; }
.markdown-preview[data-use-github-style] span.align-right { display: block; overflow: hidden; clear: both; }
.markdown-preview[data-use-github-style] span.align-right > span { display: block; margin: 13px 0px 0px; overflow: hidden; text-align: right; }
.markdown-preview[data-use-github-style] span.align-right span img { margin: 0px; text-align: right; }
.markdown-preview[data-use-github-style] span.float-left { display: block; float: left; margin-right: 13px; overflow: hidden; }
.markdown-preview[data-use-github-style] span.float-left span { margin: 13px 0px 0px; }
.markdown-preview[data-use-github-style] span.float-right { display: block; float: right; margin-left: 13px; overflow: hidden; }
.markdown-preview[data-use-github-style] span.float-right > span { display: block; margin: 13px auto 0px; overflow: hidden; text-align: right; }
.markdown-preview[data-use-github-style] code, .markdown-preview[data-use-github-style] tt { padding: 0.2em 0px; margin: 0px; font-size: 85%; background-color: rgba(0, 0, 0, 0.04); border-radius: 3px; }
.markdown-preview[data-use-github-style] code::before, .markdown-preview[data-use-github-style] tt::before, .markdown-preview[data-use-github-style] code::after, .markdown-preview[data-use-github-style] tt::after { letter-spacing: -0.2em; content: " "; }
.markdown-preview[data-use-github-style] code br, .markdown-preview[data-use-github-style] tt br { display: none; }
.markdown-preview[data-use-github-style] del code { text-decoration: inherit; }
.markdown-preview[data-use-github-style] pre > code { padding: 0px; margin: 0px; font-size: 100%; word-break: normal; white-space: pre; background: transparent; border: 0px; }
.markdown-preview[data-use-github-style] .highlight { margin-bottom: 16px; }
.markdown-preview[data-use-github-style] .highlight pre, .markdown-preview[data-use-github-style] pre { padding: 16px; overflow: auto; font-size: 85%; line-height: 1.45; background-color: rgb(247, 247, 247); border-radius: 3px; }
.markdown-preview[data-use-github-style] .highlight pre { margin-bottom: 0px; word-break: normal; }
.markdown-preview[data-use-github-style] pre { overflow-wrap: normal; }
.markdown-preview[data-use-github-style] pre code, .markdown-preview[data-use-github-style] pre tt { display: inline; max-width: initial; padding: 0px; margin: 0px; overflow: initial; line-height: inherit; overflow-wrap: normal; background-color: transparent; border: 0px; }
.markdown-preview[data-use-github-style] pre code::before, .markdown-preview[data-use-github-style] pre tt::before, .markdown-preview[data-use-github-style] pre code::after, .markdown-preview[data-use-github-style] pre tt::after { content: normal; }
.markdown-preview[data-use-github-style] kbd { display: inline-block; padding: 3px 5px; font-size: 11px; line-height: 10px; color: rgb(85, 85, 85); vertical-align: middle; background-color: rgb(252, 252, 252); border-width: 1px; border-style: solid; border-color: rgb(204, 204, 204) rgb(204, 204, 204) rgb(187, 187, 187); border-image: initial; border-radius: 3px; box-shadow: rgb(187, 187, 187) 0px -1px 0px inset; }
.markdown-preview[data-use-github-style] a { color: rgb(51, 122, 183); }
.markdown-preview[data-use-github-style] code { color: inherit; }
.markdown-preview[data-use-github-style] pre.editor-colors { padding: 0.8em 1em; margin-bottom: 1em; font-size: 0.85em; border-radius: 4px; overflow: auto; }
.markdown-preview pre.editor-colors { user-select: auto; }
.scrollbars-visible-always .markdown-preview pre.editor-colors .vertical-scrollbar, .scrollbars-visible-always .markdown-preview pre.editor-colors .horizontal-scrollbar { visibility: hidden; }
.scrollbars-visible-always .markdown-preview pre.editor-colors:hover .vertical-scrollbar, .scrollbars-visible-always .markdown-preview pre.editor-colors:hover .horizontal-scrollbar { visibility: visible; }
.markdown-preview .task-list-item input[type="checkbox"] { position: absolute; margin: 0.25em 0px 0px -1.4em; }
.markdown-preview .task-list-item { list-style-type: none; }
.bracket-matcher .region {
  border-bottom: 1px dotted lime;
  position: absolute;
}
.line-number.bracket-matcher.bracket-matcher {
  color: #383a42;
  background-color: #e5e5e6;
}

.spell-check-misspelling .region {
  border-bottom: 2px dotted rgba(255, 51, 51, 0.75);
}
.spell-check-corrections {
  width: 25em !important;
}

pre.editor-colors {
  background-color: #fafafa;
  color: #383a42;
}
pre.editor-colors .line.cursor-line {
  background-color: rgba(56, 58, 66, 0.05);
}
pre.editor-colors .invisible {
  color: #383a42;
}
pre.editor-colors .cursor {
  border-left: 2px solid #526fff;
}
pre.editor-colors .selection .region {
  background-color: #e5e5e6;
}
pre.editor-colors .bracket-matcher .region {
  border-bottom: 1px solid #526fff;
  box-sizing: border-box;
}
pre.editor-colors .invisible-character {
  color: rgba(56, 58, 66, 0.2);
}
pre.editor-colors .indent-guide {
  color: rgba(56, 58, 66, 0.2);
}
pre.editor-colors .wrap-guide {
  background-color: rgba(56, 58, 66, 0.2);
}
pre.editor-colors .find-result .region.region.region,
pre.editor-colors .current-result .region.region.region {
  border-radius: 2px;
  background-color: rgba(82, 111, 255, 0.2);
  transition: border-color 0.4s;
}
pre.editor-colors .find-result .region.region.region {
  border: 2px solid transparent;
}
pre.editor-colors .current-result .region.region.region {
  border: 2px solid #526fff;
  transition-duration: .1s;
}
pre.editor-colors .gutter .line-number {
  color: #9d9d9f;
  -webkit-font-smoothing: antialiased;
}
pre.editor-colors .gutter .line-number.cursor-line {
  color: #383a42;
  background-color: #e5e5e6;
}
pre.editor-colors .gutter .line-number.cursor-line-no-selection {
  background-color: transparent;
}
pre.editor-colors .gutter .line-number .icon-right {
  color: #383a42;
}
pre.editor-colors .gutter:not(.git-diff-icon) .line-number.git-line-removed.git-line-removed::before {
  bottom: -3px;
}
pre.editor-colors .gutter:not(.git-diff-icon) .line-number.git-line-removed::after {
  content: "";
  position: absolute;
  left: 0px;
  bottom: 0px;
  width: 25px;
  border-bottom: 1px dotted rgba(255, 20, 20, 0.5);
  pointer-events: none;
}
pre.editor-colors .gutter .line-number.folded,
pre.editor-colors .gutter .line-number:after,
pre.editor-colors .fold-marker:after {
  color: #383a42;
}
.syntax--comment {
  color: #a0a1a7;
  font-style: italic;
}
.syntax--comment .syntax--markup.syntax--link {
  color: #a0a1a7;
}
.syntax--entity.syntax--name.syntax--type {
  color: #c18401;
}
.syntax--entity.syntax--other.syntax--inherited-class {
  color: #c18401;
}
.syntax--keyword {
  color: #a626a4;
}
.syntax--keyword.syntax--control {
  color: #a626a4;
}
.syntax--keyword.syntax--operator {
  color: #383a42;
}
.syntax--keyword.syntax--other.syntax--special-method {
  color: #4078f2;
}
.syntax--keyword.syntax--other.syntax--unit {
  color: #986801;
}
.syntax--storage {
  color: #a626a4;
}
.syntax--storage.syntax--type.syntax--annotation,
.syntax--storage.syntax--type.syntax--primitive {
  color: #a626a4;
}
.syntax--storage.syntax--modifier.syntax--package,
.syntax--storage.syntax--modifier.syntax--import {
  color: #383a42;
}
.syntax--constant {
  color: #986801;
}
.syntax--constant.syntax--variable {
  color: #986801;
}
.syntax--constant.syntax--character.syntax--escape {
  color: #0184bc;
}
.syntax--constant.syntax--numeric {
  color: #986801;
}
.syntax--constant.syntax--other.syntax--color {
  color: #0184bc;
}
.syntax--constant.syntax--other.syntax--symbol {
  color: #0184bc;
}
.syntax--variable {
  color: #e45649;
}
.syntax--variable.syntax--interpolation {
  color: #ca1243;
}
.syntax--variable.syntax--parameter {
  color: #383a42;
}
.syntax--string {
  color: #50a14f;
}
.syntax--string > .syntax--source,
.syntax--string .syntax--embedded {
  color: #383a42;
}
.syntax--string.syntax--regexp {
  color: #0184bc;
}
.syntax--string.syntax--regexp .syntax--source.syntax--ruby.syntax--embedded {
  color: #c18401;
}
.syntax--string.syntax--other.syntax--link {
  color: #e45649;
}
.syntax--punctuation.syntax--definition.syntax--comment {
  color: #a0a1a7;
}
.syntax--punctuation.syntax--definition.syntax--method-parameters,
.syntax--punctuation.syntax--definition.syntax--function-parameters,
.syntax--punctuation.syntax--definition.syntax--parameters,
.syntax--punctuation.syntax--definition.syntax--separator,
.syntax--punctuation.syntax--definition.syntax--seperator,
.syntax--punctuation.syntax--definition.syntax--array {
  color: #383a42;
}
.syntax--punctuation.syntax--definition.syntax--heading,
.syntax--punctuation.syntax--definition.syntax--identity {
  color: #4078f2;
}
.syntax--punctuation.syntax--definition.syntax--bold {
  color: #c18401;
  font-weight: bold;
}
.syntax--punctuation.syntax--definition.syntax--italic {
  color: #a626a4;
  font-style: italic;
}
.syntax--punctuation.syntax--section.syntax--embedded {
  color: #ca1243;
}
.syntax--punctuation.syntax--section.syntax--method,
.syntax--punctuation.syntax--section.syntax--class,
.syntax--punctuation.syntax--section.syntax--inner-class {
  color: #383a42;
}
.syntax--support.syntax--class {
  color: #c18401;
}
.syntax--support.syntax--type {
  color: #0184bc;
}
.syntax--support.syntax--function {
  color: #0184bc;
}
.syntax--support.syntax--function.syntax--any-method {
  color: #4078f2;
}
.syntax--entity.syntax--name.syntax--function {
  color: #4078f2;
}
.syntax--entity.syntax--name.syntax--class,
.syntax--entity.syntax--name.syntax--type.syntax--class {
  color: #c18401;
}
.syntax--entity.syntax--name.syntax--section {
  color: #4078f2;
}
.syntax--entity.syntax--name.syntax--tag {
  color: #e45649;
}
.syntax--entity.syntax--other.syntax--attribute-name {
  color: #986801;
}
.syntax--entity.syntax--other.syntax--attribute-name.syntax--id {
  color: #4078f2;
}
.syntax--meta.syntax--class {
  color: #c18401;
}
.syntax--meta.syntax--class.syntax--body {
  color: #383a42;
}
.syntax--meta.syntax--method-call,
.syntax--meta.syntax--method {
  color: #383a42;
}
.syntax--meta.syntax--definition.syntax--variable {
  color: #e45649;
}
.syntax--meta.syntax--link {
  color: #986801;
}
.syntax--meta.syntax--require {
  color: #4078f2;
}
.syntax--meta.syntax--selector {
  color: #a626a4;
}
.syntax--meta.syntax--separator {
  color: #383a42;
}
.syntax--meta.syntax--tag {
  color: #383a42;
}
.syntax--underline {
  text-decoration: underline;
}
.syntax--none {
  color: #383a42;
}
.syntax--invalid.syntax--deprecated {
  color: #000000 !important;
  background-color: #f2a60d !important;
}
.syntax--invalid.syntax--illegal {
  color: white !important;
  background-color: #ff1414 !important;
}
.syntax--markup.syntax--bold {
  color: #986801;
  font-weight: bold;
}
.syntax--markup.syntax--changed {
  color: #a626a4;
}
.syntax--markup.syntax--deleted {
  color: #e45649;
}
.syntax--markup.syntax--italic {
  color: #a626a4;
  font-style: italic;
}
.syntax--markup.syntax--heading {
  color: #e45649;
}
.syntax--markup.syntax--heading .syntax--punctuation.syntax--definition.syntax--heading {
  color: #4078f2;
}
.syntax--markup.syntax--link {
  color: #0184bc;
}
.syntax--markup.syntax--inserted {
  color: #50a14f;
}
.syntax--markup.syntax--quote {
  color: #986801;
}
.syntax--markup.syntax--raw {
  color: #50a14f;
}
.syntax--source.syntax--c .syntax--keyword.syntax--operator {
  color: #a626a4;
}
.syntax--source.syntax--cpp .syntax--keyword.syntax--operator {
  color: #a626a4;
}
.syntax--source.syntax--cs .syntax--keyword.syntax--operator {
  color: #a626a4;
}
.syntax--source.syntax--css .syntax--property-name,
.syntax--source.syntax--css .syntax--property-value {
  color: #696c77;
}
.syntax--source.syntax--css .syntax--property-name.syntax--support,
.syntax--source.syntax--css .syntax--property-value.syntax--support {
  color: #383a42;
}
.syntax--source.syntax--elixir .syntax--source.syntax--embedded.syntax--source {
  color: #383a42;
}
.syntax--source.syntax--elixir .syntax--constant.syntax--language,
.syntax--source.syntax--elixir .syntax--constant.syntax--numeric,
.syntax--source.syntax--elixir .syntax--constant.syntax--definition {
  color: #4078f2;
}
.syntax--source.syntax--elixir .syntax--variable.syntax--definition,
.syntax--source.syntax--elixir .syntax--variable.syntax--anonymous {
  color: #a626a4;
}
.syntax--source.syntax--elixir .syntax--parameter.syntax--variable.syntax--function {
  color: #986801;
  font-style: italic;
}
.syntax--source.syntax--elixir .syntax--quoted {
  color: #50a14f;
}
.syntax--source.syntax--elixir .syntax--keyword.syntax--special-method,
.syntax--source.syntax--elixir .syntax--embedded.syntax--section,
.syntax--source.syntax--elixir .syntax--embedded.syntax--source.syntax--empty {
  color: #e45649;
}
.syntax--source.syntax--elixir .syntax--readwrite.syntax--module .syntax--punctuation {
  color: #e45649;
}
.syntax--source.syntax--elixir .syntax--regexp.syntax--section,
.syntax--source.syntax--elixir .syntax--regexp.syntax--string {
  color: #ca1243;
}
.syntax--source.syntax--elixir .syntax--separator,
.syntax--source.syntax--elixir .syntax--keyword.syntax--operator {
  color: #986801;
}
.syntax--source.syntax--elixir .syntax--variable.syntax--constant {
  color: #c18401;
}
.syntax--source.syntax--elixir .syntax--array,
.syntax--source.syntax--elixir .syntax--scope,
.syntax--source.syntax--elixir .syntax--section {
  color: #696c77;
}
.syntax--source.syntax--gfm .syntax--markup {
  -webkit-font-smoothing: auto;
}
.syntax--source.syntax--gfm .syntax--link .syntax--entity {
  color: #4078f2;
}
.syntax--source.syntax--go .syntax--storage.syntax--type.syntax--string {
  color: #a626a4;
}
.syntax--source.syntax--ini .syntax--keyword.syntax--other.syntax--definition.syntax--ini {
  color: #e45649;
}
.syntax--source.syntax--java .syntax--storage.syntax--modifier.syntax--import {
  color: #c18401;
}
.syntax--source.syntax--java .syntax--storage.syntax--type {
  color: #c18401;
}
.syntax--source.syntax--java .syntax--keyword.syntax--operator.syntax--instanceof {
  color: #a626a4;
}
.syntax--source.syntax--java-properties .syntax--meta.syntax--key-pair {
  color: #e45649;
}
.syntax--source.syntax--java-properties .syntax--meta.syntax--key-pair > .syntax--punctuation {
  color: #383a42;
}
.syntax--source.syntax--js .syntax--keyword.syntax--operator {
  color: #0184bc;
}
.syntax--source.syntax--js .syntax--keyword.syntax--operator.syntax--delete,
.syntax--source.syntax--js .syntax--keyword.syntax--operator.syntax--in,
.syntax--source.syntax--js .syntax--keyword.syntax--operator.syntax--of,
.syntax--source.syntax--js .syntax--keyword.syntax--operator.syntax--instanceof,
.syntax--source.syntax--js .syntax--keyword.syntax--operator.syntax--new,
.syntax--source.syntax--js .syntax--keyword.syntax--operator.syntax--typeof,
.syntax--source.syntax--js .syntax--keyword.syntax--operator.syntax--void {
  color: #a626a4;
}
.syntax--source.syntax--ts .syntax--keyword.syntax--operator {
  color: #0184bc;
}
.syntax--source.syntax--flow .syntax--keyword.syntax--operator {
  color: #0184bc;
}
.syntax--source.syntax--json .syntax--meta.syntax--structure.syntax--dictionary.syntax--json > .syntax--string.syntax--quoted.syntax--json {
  color: #e45649;
}
.syntax--source.syntax--json .syntax--meta.syntax--structure.syntax--dictionary.syntax--json > .syntax--string.syntax--quoted.syntax--json > .syntax--punctuation.syntax--string {
  color: #e45649;
}
.syntax--source.syntax--json .syntax--meta.syntax--structure.syntax--dictionary.syntax--json > .syntax--value.syntax--json > .syntax--string.syntax--quoted.syntax--json,
.syntax--source.syntax--json .syntax--meta.syntax--structure.syntax--array.syntax--json > .syntax--value.syntax--json > .syntax--string.syntax--quoted.syntax--json,
.syntax--source.syntax--json .syntax--meta.syntax--structure.syntax--dictionary.syntax--json > .syntax--value.syntax--json > .syntax--string.syntax--quoted.syntax--json > .syntax--punctuation,
.syntax--source.syntax--json .syntax--meta.syntax--structure.syntax--array.syntax--json > .syntax--value.syntax--json > .syntax--string.syntax--quoted.syntax--json > .syntax--punctuation {
  color: #50a14f;
}
.syntax--source.syntax--json .syntax--meta.syntax--structure.syntax--dictionary.syntax--json > .syntax--constant.syntax--language.syntax--json,
.syntax--source.syntax--json .syntax--meta.syntax--structure.syntax--array.syntax--json > .syntax--constant.syntax--language.syntax--json {
  color: #0184bc;
}
.syntax--ng.syntax--interpolation {
  color: #e45649;
}
.syntax--ng.syntax--interpolation.syntax--begin,
.syntax--ng.syntax--interpolation.syntax--end {
  color: #4078f2;
}
.syntax--ng.syntax--interpolation .syntax--function {
  color: #e45649;
}
.syntax--ng.syntax--interpolation .syntax--function.syntax--begin,
.syntax--ng.syntax--interpolation .syntax--function.syntax--end {
  color: #4078f2;
}
.syntax--ng.syntax--interpolation .syntax--bool {
  color: #986801;
}
.syntax--ng.syntax--interpolation .syntax--bracket {
  color: #383a42;
}
.syntax--ng.syntax--pipe,
.syntax--ng.syntax--operator {
  color: #383a42;
}
.syntax--ng.syntax--tag {
  color: #0184bc;
}
.syntax--ng.syntax--attribute-with-value .syntax--attribute-name {
  color: #c18401;
}
.syntax--ng.syntax--attribute-with-value .syntax--string {
  color: #a626a4;
}
.syntax--ng.syntax--attribute-with-value .syntax--string.syntax--begin,
.syntax--ng.syntax--attribute-with-value .syntax--string.syntax--end {
  color: #383a42;
}
.syntax--source.syntax--ruby .syntax--constant.syntax--other.syntax--symbol > .syntax--punctuation {
  color: inherit;
}
.syntax--source.syntax--php .syntax--class.syntax--bracket {
  color: #383a42;
}
.syntax--source.syntax--python .syntax--keyword.syntax--operator.syntax--logical.syntax--python {
  color: #a626a4;
}
.syntax--source.syntax--python .syntax--variable.syntax--parameter {
  color: #986801;
}
</style>
        </head>
        <body class='markdown-preview' data-use-github-style><h1 id="seqapipop-analyses-from-fastq-files-to-vcf-for-870-samples">SeqApiPop analyses: from fastq files to vcf for 870 samples</h1>


<ul>
<li><a href="#1-introduction">1. Introduction</a></li>
<li><a href="#2-from-fastq-files-to-gvcf-files">2. From fastq files to gvcf files</a><ul>
<li><a href="#21-reference-genome">2.1. Reference genome</a></li>
<li><a href="#22-fastq-files">2.2. fastq files</a><ul>
<li><a href="#221-availability">2.2.1. Availability</a></li>
<li><a href="#222-specific-cases">2.2.2. Specific cases</a><ul>
<li><a href="#2221-sample-br1-has-two-library-tags">2.2.2.1. Sample BR1 has two library tags</a></li>
<li><a href="#2222-samples-sequenced-in-multiple-lanes-from-one-library">2.2.2.2. Samples sequenced in multiple lanes from one library</a></li>
</ul>
</li>
<li><a href="#223-preparing-for-mapping">2.2.3. Preparing for mapping</a></li>
</ul>
</li>
<li><a href="#23-mapping-calling-script-mapseqapipophav31bash">2.3. Mapping: Calling script: Map_seqapipop_HAV3_1.bash</a><ul>
<li><a href="#231-mapseqapipophav31bash">2.3.1. Map_seqapipop_HAV3_1.bash</a></li>
<li><a href="#232-ploidy-variable">2.3.2. PLOIDY variable</a></li>
<li><a href="#233-n-variable">2.3.3. N variable</a></li>
<li><a href="#234-samplefile-variable">2.3.4. SAMPLE_FILE variable</a></li>
<li><a href="#235-out-variable">2.3.5. OUT variable</a></li>
<li><a href="#236-ref-variable">2.3.6. REF variable</a></li>
<li><a href="#237-chromosome-variable">2.3.7. CHROMOSOME variable</a></li>
<li><a href="#238-unknown-variable">2.3.8. UNKNOWN variable</a></li>
<li><a href="#239-script-variable">2.3.9. SCRIPT variable</a></li>
<li><a href="#2310-output-files">2.3.10. Output files</a></li>
</ul>
</li>
<li><a href="#24-merging-bams">2.4. Merging bams</a><ul>
<li><a href="#241-preparing-files-and-scripts">2.4.1. Preparing files and scripts</a></li>
<li><a href="#242-list-of-samples-to-merge">2.4.2. List of samples to merge</a></li>
<li><a href="#243-merging-calling-script-callmergingsh">2.4.3. Merging: Calling script callMerging.sh</a><ul>
<li><a href="#2431-callmergingsh">2.4.3.1. callMerging.sh</a></li>
<li><a href="#2432-samplefile-variable">2.4.3.2. SAMPLE_FILE variable</a></li>
<li><a href="#2432-startpoint-variable">2.4.3.2. STARTPOINT variable</a></li>
<li><a href="#2433-other-variables">2.4.3.3. Other variables</a></li>
</ul>
</li>
<li><a href="#2434-output-files">2.4.3.4 Output files</a></li>
<li><a href="#244-bqsr-and-calling-on-the-merged-bams">2.4.4. BQSR and calling on the merged bams</a><ul>
<li><a href="#2441-bootstrapcallbash-script">2.4.4.1. Bootstrap_Call.bash script</a></li>
<li><a href="#2442-the-samplefile-variable">2.4.4.2. The SAMPLE_FILE variable</a></li>
<li><a href="#2442-other-variables">2.4.4.2. Other variables</a></li>
</ul>
</li>
<li><a href="#2443-output-files">2.4.4.3. Output files</a></li>
<li><a href="#2444-same-samples-under-different-names">2.4.4.4. Same samples under different names.</a></li>
</ul>
</li>
<li><a href="#25-controlling-that-all-gvcf-files-are-complete">2.5 controlling that all gvcf files are complete</a><ul>
<li><a href="#251-script-controlvcfsbash">2.5.1 Script controlVcfs.bash</a></li>
<li><a href="#252-check-all-went-to-the-end">2.5.2. Check all went to the end</a></li>
</ul>
</li>
</ul>
</li>
<li><a href="#3-making-the-large-vcf-file-with-all-870-samples">3. Making the large vcf file with all 870 samples</a><ul>
<li><a href="#31-combine-gvcf-files">3.1 Combine gvcf files</a></li>
<li><a href="#311-generate-list-of-all-samples-for-the-combine-script">3.1.1. Generate list of all samples for the combine script</a></li>
<li><a href="#312-write-the-combining-scripts">3.1.2 Write the combining scripts</a></li>
<li><a href="#313-run-the-combine-script-with-combinegvcfshav31lanceslurmbash">3.1.3 Run the combine script with combineGVCFsHAV3_1_Lance_slurm.bash</a></li>
<li><a href="#314-output">3.1.4. Output</a></li>
<li><a href="#315-check-all-went-to-the-end">3.1.5. Check all went to the end</a><ul>
<li><a href="#3151-example-for-mitochondrial-dna">3.1.5.1. Example for mitochondrial DNA</a></li>
<li><a href="#3152-for-all-chromosomes">3.1.5.2. For all chromosomes:</a></li>
</ul>
</li>
<li><a href="#32-genotype-across-the-combinegvcfs">3.2. Genotype across the CombineGVCFs</a><ul>
<li><a href="#321-genotype">3.2.1. Genotype</a></li>
<li><a href="#322-check-output">3.2.2. Check output</a></li>
</ul>
</li>
<li><a href="#33-concatenate-vcf-files">3.3 Concatenate vcf files</a></li>
<li><a href="#34-count-variants-per-chromosomes">3.4. Count variants per chromosomes</a></li>
<li><a href="#35-retain-only-snps">3.5 Retain only SNPs</a></li>
</ul>
</li>
</ul>


<h2 id="1-introduction">1. Introduction</h2>
<p>This document presents the analyses performed for obtaining the results presented in the paper <em>SeqApiPop</em> by Wragg et al (2020). Scripts will have to be adapted to your local environment (SGE vs Slurm cluster; paths...)</p>
<p>The main scripts are in <a href="ScriptsMappingCalling">ScriptsMappingCalling</a>.</p>
<p>Short scripts are directly described in the text.</p>
<p>Versions of the software used:</p>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">more</span> program_module</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#%Module1.0###############################################################</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">##</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">module</span> load bioinfo/bwa-0.7.15</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">module</span> load bioinfo/samtools-1.8</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">module</span> load bioinfo/bedtools-2.27.1</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">module</span> load bioinfo/tabix-0.2.5</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">module</span> load bioinfo/vcftools-0.1.15</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">module</span> load bioinfo/bcftools-1.6</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">module</span> load bioinfo/picard-2.18.2</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">module</span> load bioinfo/gatk-4.1.2.0</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">module</span> load bioinfo/plink-v1.90b5.3</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">module</span> load system/Python-3.6.3</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">module</span> load system/Java8</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">module</span> load system/R-3.4.3</span></div></pre>
<h2 id="2-from-fastq-files-to-gvcf-files">2. From fastq files to gvcf files</h2>
<h3 id="21-reference-genome">2.1. Reference genome</h3>
<p>The reference genome used is the <a href="https://www.ncbi.nlm.nih.gov/assembly/GCA_003254395.2">Amel_HAv3.1</a>, having Genbank and RefSeq assembly accessions GenBank assembly accessions
    GCA_003254395.2 and GCF_003254395.2.</p>
<h3 id="22-fastq-files">2.2. fastq files</h3>
<h4 id="221-availability">2.2.1. Availability</h4>
<p>The individual fastq files for the population analysis were produced by Illumlina sequencing machines. These raw sequence reads are available for download at NCBI within the SeqApiPop project <a href="https://www.ncbi.nlm.nih.gov/bioproject/PRJNA311274">PRJNA311274</a> and SRA runs of accession <a href="https://trace.ncbi.nlm.nih.gov/Traces/study/?acc=SRP069814&amp;o=acc_s%3Aa">SRP069814</a>.</p>
<p>Naming of sequence files is in the form "SampleId_LibraryTag_SequencingLane_Read1or2.fastq.gz". For instance: OUE9_GTCCGC_L008_R1.fastq.gz</p>
<h4 id="222-specific-cases">2.2.2. Specific cases</h4>
<h5 id="2221-sample-br1-has-two-library-tags">2.2.2.1. Sample BR1 has two library tags</h5>
<ul>
<li>BR1_ATGTCA_L003</li>
<li>BR1_CCGTCC_L001<ul>
<li>These were considered as two different samples and renamed BR1A and BR1B</li>
</ul>
</li>
</ul>
<h5 id="2222-samples-sequenced-in-multiple-lanes-from-one-library">2.2.2.2. Samples sequenced in multiple lanes from one library</h5>
<p>For samples having low depth of coverage in a first run, the library was re-loaded for sequencing in a second run.
Samples sequenced in multiple lanes were on two lanes, except OUE8 that was on 3 lanes. Mapping was done separately and the bam files were merged after the mapping.</p>
<h4 id="223-preparing-for-mapping">2.2.3. Preparing for mapping</h4>
<p>For each sample/run, there is one *_R1.fastq.gz and one *_R2.fastq.gz file with the forward and reverse reads. In our case, these are all in the directory <em>FastqFromNG6</em>.</p>
<h3 id="23-mapping-calling-script-map_seqapipop_hav3_1bash">2.3. Mapping: Calling script: Map_seqapipop_HAV3_1.bash</h3>
<p>The mapping scripts perform mapping with BWA, marking duplicate reads with Picard, GATK BQSR (Base Quality Score Recalibration) and variant calling with GATK HaplotypeCaller, thus producing individual *.gvcf files.</p>
<h4 id="231-map_seqapipop_hav3_1bash">2.3.1. Map_seqapipop_HAV3_1.bash</h4>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#!/bin/bash</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block"># Usage:        Map_seqapipop_HAV3_1.bash</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--variable syntax--other syntax--member">N</span>=1</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--variable syntax--other syntax--member">PLOIDY</span>=2</span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#------------------------------------------------------------------</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#EDIT FOR ACCESS TO SEQUENCE FILES</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block"># First download links from NG6 to genotoul and update the path here</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--variable syntax--other syntax--member">SAMPLE_FILE</span>=/work/project/cytogen/Alain/seqapipopOnHAV3_1/O.list</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--variable syntax--other syntax--member">OUT</span>=/work/project/cytogen/Alain/seqapipopOnHAV3_1/Map_O</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--variable syntax--other syntax--member">REF</span>=/home/gencel/vignal/save/Genomes/Abeille/HAv3_1_indexes/GCF_003254395.2_Amel_HAv3.1_genomic.fna</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--variable syntax--other syntax--member">CHROMOSOME</span>=/home/gencel/vignal/save/Genomes/Abeille/HAv3_1_indexes/HAv3_1_Chromosomes.list</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--variable syntax--other syntax--member">UNKNOWN</span>=/home/gencel/vignal/save/Genomes/Abeille/HAv3_1_indexes/HAv3_1_Unknown.list</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--variable syntax--other syntax--member">SCRIPT</span>=/genphyse/cytogen/seqapipop/ScriptsHAV3Mapping/mapping_calling</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#------------------------------------------------------------------</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">mkdir</span> <span class="syntax--entity syntax--other syntax--attribute-name">-p</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OUT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/logs/mapping    <span class="syntax--comment syntax--block"># /work/ktabetaoul/Project_seqapipop/RESULTATS/logs/mapping</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function"><span class="syntax--support syntax--function">cd</span></span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OUT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#EDIT FOR mapping.sh</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block"># This is the loop to send each sample to the cluster to be mapped</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--keyword syntax--control">while</span> <span class="syntax--entity syntax--name syntax--function">read</span> line</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--keyword syntax--control">do</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--variable syntax--other syntax--member">ID</span>=`<span class="syntax--entity syntax--name syntax--function">basename</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">line</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>`   <span class="syntax--comment syntax--block"># give the name of the sample-population to map</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--variable syntax--other syntax--member">IN</span>=`<span class="syntax--entity syntax--name syntax--function">dirname</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">line</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>`    <span class="syntax--comment syntax--block"># give the path where find the sample fastq</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--entity syntax--name syntax--function">mkdir</span> <span class="syntax--entity syntax--other syntax--attribute-name">-p</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OUT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">ID</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--entity syntax--name syntax--function">sbatch</span> <span class="syntax--entity syntax--other syntax--attribute-name">--cpus-per-task=1</span> <span class="syntax--entity syntax--other syntax--attribute-name">--mem-per-cpu=5G</span> \</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">                </span><span class="syntax--entity syntax--other syntax--attribute-name">-J</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">ID</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>_mapping <span class="syntax--entity syntax--other syntax--attribute-name">-o</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OUT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/logs/<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">ID</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>_mapping.o <span class="syntax--entity syntax--other syntax--attribute-name">-e</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OUT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/logs/<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">ID</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>_mapping.e \</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">                </span><span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">SCRIPT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/mappingAV_2019_Dec.sh <span class="syntax--entity syntax--other syntax--attribute-name">-s</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">ID</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span> <span class="syntax--entity syntax--other syntax--attribute-name">-i</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">IN</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span> <span class="syntax--entity syntax--other syntax--attribute-name">-o</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OUT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">ID</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span> <span class="syntax--entity syntax--other syntax--attribute-name">-p</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">PLOIDY</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span> <span class="syntax--entity syntax--other syntax--attribute-name">-n</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">N</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span> <span class="syntax--entity syntax--other syntax--attribute-name">-e</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OUT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/logs <span class="syntax--entity syntax--other syntax--attribute-name">-R</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">REF</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span> <span class="syntax--entity syntax--other syntax--attribute-name">-C</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">CHROMOSOME</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span> <span class="syntax--entity syntax--other syntax--attribute-name">-U</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">UNKNOWN</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--keyword syntax--control">done</span> <span class="syntax--keyword syntax--operator">&lt;</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">SAMPLE_FILE</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block"># end of file</span></span></div></pre>
<h4 id="232-ploidy-variable">2.3.2. PLOIDY variable</h4>
<p>The variable PLOIDY used for specifying the HaplotypeCaller model to call variants is set here to POILDY=2 although the drones sequenced are haploid. This was done following first tests with the haploid model (PLOIDY=1), showing this will lead to false genotype calling in repeated sequences, especially VNTRs. Indeed, these will cause reads from different copies of a repeated sequence to map to a same position, creating false positive SNPs. In the haploid model, these will be given arbitrarily one of the possible alleles, with a high GQ (Genotype Quality), whereas in the diploid model, they will be caled as heterozygotes. These heterozygote positions can then easily be filterd out in subsequent steps.</p>
<h4 id="233-n-variable">2.3.3. N variable</h4>
<p>Was introduced for eventual sequencing of DNA pools of individuals. Here, we perform individual sequencing. Is thus set to N=1</p>
<h4 id="234-sample_file-variable">2.3.4. SAMPLE_FILE variable</h4>
<p>Contains a list of samples to process, in the form:</p>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function"><span class="syntax--punctuation syntax--section syntax--embedded">$</span> <span class="syntax--variable syntax--other syntax--member">head</span></span> <span class="syntax--entity syntax--other syntax--attribute-name">-3</span> O.list</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">/genphyse/cytogen/seqapipop/FastqFromNG6/OUE10_GTGAAA_L003</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">/genphyse/cytogen/seqapipop/FastqFromNG6/OUE10_GTGAAA_L008</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">/genphyse/cytogen/seqapipop/FastqFromNG6/OUE11_GTGGCC_L003</span></span></div></pre>
<h4 id="235-out-variable">2.3.5. OUT variable</h4>
<p>Directory for the output.</p>
<h4 id="236-ref-variable">2.3.6. REF variable</h4>
<p>Path to the reference genome, indexed for BWA</p>
<h4 id="237-chromosome-variable">2.3.7. CHROMOSOME variable</h4>
<p>Path to the list of chromosomes and their length. List in the form:</p>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function"><span class="syntax--punctuation syntax--section syntax--embedded">$</span> <span class="syntax--variable syntax--other syntax--member">more</span></span> HAv3_1_Chromosomes.list</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037638.1</span>     27754200</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037639.1</span>     16089512</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037640.1</span>     13619445</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037641.1</span>     13404451</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037642.1</span>     13896941</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037643.1</span>     17789102</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037644.1</span>     14198698</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037645.1</span>     12717210</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037646.1</span>     12354651</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037647.1</span>     12360052</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037648.1</span>     16352600</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037649.1</span>     11514234</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037650.1</span>     11279722</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037651.1</span>     10670842</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037652.1</span>     9534514</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037653.1</span>     7238532</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_001566.1</span>     16343</span></div></pre>
<h4 id="238-unknown-variable">2.3.8. UNKNOWN variable</h4>
<p>Path to the list of unknown and/or unplaced scaffolds and their length. List in the form:</p>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">head</span> <span class="syntax--entity syntax--other syntax--attribute-name">-3</span> HAv3_1_Unknown.list</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NW_020555788.1</span>  3988</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NW_020555789.1</span>  24043</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NW_020555790.1</span>  23881</span></div></pre>
<h4 id="239-script-variable">2.3.9. SCRIPT variable</h4>
<p>Path to the directory containing the mapping, BQSR and calling scripts</p>
<ul>
<li>mappingAV_2019_Dec.sh : does the mapping and duplicate reads detection. Called by Map_seqapipop_HAV3_1.bash</li>
<li>bootstrapingAV_2019_Dec.sh : does the BQSR. Called by mappingAV_2019_Dec.sh</li>
<li>callingAV_2019_Dec.sh : does the HaplotypeCaller variant calling. Called by mappingAV_2019_Dec.sh</li>
</ul>
<h4 id="2310-output-files">2.3.10. Output files</h4>
<p>For each sample, output files will be in a directory with the sample name (SAMPLE_NAME) in the form "SampleId_LibraryTag_SequencingLane", each containing 4 directories, with the bam and gvcf files:</p>
<ul>
<li>SAMPLE_NAME_sort.bam (before BQSR)</li>
<li>SAMPLE_NAME.bam (after BQSR)</li>
<li>SAMPLE_NAME.g.vcf.gz</li>
<li>SAMPLE_NAME_dedup.metrics</li>
</ul>
<p>For instance:</p>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function"><span class="syntax--punctuation syntax--section syntax--embedded">$</span> <span class="syntax--variable syntax--other syntax--member">ls</span></span> OUE9_GTCCGC_L008/*/</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">OUE9_GTCCGC_L008/bootstraping/:</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">OUE9_GTCCGC_L008.bam</span>  OUE9_GTCCGC_L008.bam.bai  OUE9_GTCCGC_L008_bam.list</span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">OUE9_GTCCGC_L008/calling/:</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">OUE9_GTCCGC_L008.g.vcf.gz</span>  OUE9_GTCCGC_L008.g.vcf.gz.tbi</span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">OUE9_GTCCGC_L008/mapping/:</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">OUE9_GTCCGC_L008_sort.bam</span>  OUE9_GTCCGC_L008_sort.bam.bai</span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">OUE9_GTCCGC_L008/metrics/:</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">OUE9_GTCCGC_L008_dedup.metrics</span></span></div></pre>
<h3 id="24-merging-bams">2.4. Merging bams</h3>
<h4 id="241-preparing-files-and-scripts">2.4.1. Preparing files and scripts</h4>
<p>For each sample/run, there is one *_R1.fastq.gz and one *_R2.fastq.gz file with the forward and reverse reads. These are al in the directory <em>FastqFromNG6</em>.</p>
<h4 id="242-list-of-samples-to-merge">2.4.2. List of samples to merge</h4>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">ls</span> FastqFromNG6/*_R1.fastq.gz <span class="syntax--keyword syntax--operator">|</span> <span class="syntax--entity syntax--name syntax--function">awk</span> <span class="syntax--string">'BEGIN{FS="/"}{print $6}'</span> <span class="syntax--keyword syntax--operator">|</span> <span class="syntax--entity syntax--name syntax--function">awk</span> <span class="syntax--string">'BEGIN{FS=OFS="_"}{print $1,$2}'</span> <span class="syntax--keyword syntax--operator">|</span> <span class="syntax--entity syntax--name syntax--function">sort</span> <span class="syntax--keyword syntax--operator">|</span> <span class="syntax--entity syntax--name syntax--function">uniq</span> <span class="syntax--entity syntax--other syntax--attribute-name">-c</span> <span class="syntax--keyword syntax--operator">|</span> <span class="syntax--entity syntax--name syntax--function">sort</span> <span class="syntax--keyword syntax--operator">&gt;</span> ToMerge.list</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">awk</span> <span class="syntax--string">'$1 &gt; 1 {print $2}'</span> ToMerge.list <span class="syntax--keyword syntax--operator">|</span> <span class="syntax--entity syntax--name syntax--function">awk</span> <span class="syntax--string">'BEGIN{FS="_"}{print $1}'</span> <span class="syntax--keyword syntax--operator">&gt;</span> SamplesToMerge.list</span></div></pre>
<p>The list is in the form:</p>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function"><span class="syntax--punctuation syntax--section syntax--embedded">$</span> <span class="syntax--variable syntax--other syntax--member">head</span></span> <span class="syntax--entity syntax--other syntax--attribute-name">-3</span> SamplesToMerge.list</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">BER15</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">BER16</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">BER17</span></span></div></pre>
<h4 id="243-merging-calling-script-callmergingsh">2.4.3. Merging: Calling script callMerging.sh</h4>
<p>The merging script will take a list of samples to merge, merge the corresponding bam files together and re-evaluate duplicate reads. Output will be merged bams in a directory structure identical to the standard mapping, except for the SequencinghLane identifier in "SampleId_LibraryTag_SequencingLane" replaced by "merged": ~/OUE9_GTCCGC_merged/mapping/OUE9_GTCCGC_merged_sort.bam. These will have to be run through the BQSR and calling processes with the script Bootstrap_Call.bash.</p>
<h5 id="2431-callmergingsh">2.4.3.1. callMerging.sh</h5>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#!/bin/bash</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block"># Usage:        callMerging.sh</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#------------------------------------------------------------------</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#EDIT FOR ACCESS TO BAM FILES</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--variable syntax--other syntax--member">SAMPLE_FILE</span>=/work/project/cytogen/Alain/seqapipopOnHAV3_1/Merging/SamplesToMerge.list  <span class="syntax--comment syntax--block">#Names in the form OUE8</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--variable syntax--other syntax--member">STARTPOINT</span>=/genphyse/cytogen/seqapipop/Data/Apis-mellifera/seqapipopOnHAV3_1/*/*/mapping <span class="syntax--comment syntax--block">#Where the bam files are at the time : in sub-folders</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--variable syntax--other syntax--member">OUT</span>=/work/project/cytogen/Alain/seqapipopOnHAV3_1/Merging/MergedBams</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--variable syntax--other syntax--member">REF</span>=/home/gencel/vignal/save/Genomes/Abeille/HAv3_1_indexes/GCF_003254395.2_Amel_HAv3.1_genomic.fna</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--variable syntax--other syntax--member">CHROMOSOME</span>=/home/gencel/vignal/save/Genomes/Abeille/HAv3_1_indexes/HAv3_1_Chromosomes.list</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--variable syntax--other syntax--member">UNKNOWN</span>=/home/gencel/vignal/save/Genomes/Abeille/HAv3_1_indexes/HAv3_1_Unknown.list</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--variable syntax--other syntax--member">SCRIPT</span>=/genphyse/cytogen/seqapipop/ScriptsHAV3Mapping/mapping_calling</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#------------------------------------------------------------------</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">mkdir</span> <span class="syntax--entity syntax--other syntax--attribute-name">-p</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OUT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/logs/merging</span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--keyword syntax--control">while</span> <span class="syntax--entity syntax--name syntax--function">read</span> sample</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--keyword syntax--control">do</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--variable syntax--other syntax--member">BAM_LIST</span>=(`<span class="syntax--entity syntax--name syntax--function">ls</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">STARTPOINT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">sample</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>_*.bam`)</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--variable syntax--other syntax--member">DIR</span>=`<span class="syntax--entity syntax--name syntax--function">dirname</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">BAM_LIST</span>[0]<span class="syntax--punctuation syntax--section syntax--embedded">}</span>`</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--variable syntax--other syntax--member">IN</span>=`<span class="syntax--entity syntax--name syntax--function">basename</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">BAM_LIST</span>[0]<span class="syntax--punctuation syntax--section syntax--embedded">}</span>`</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--variable syntax--other syntax--member">SAMPLE</span>=<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">IN</span>%_*<span class="syntax--punctuation syntax--section syntax--embedded">}</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--variable syntax--other syntax--member">SAMPLE_SHORT</span>=<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">SAMPLE</span>%_*<span class="syntax--punctuation syntax--section syntax--embedded">}</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--variable syntax--other syntax--member">SAMPLE_MERGED</span>=<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">SAMPLE_SHORT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>_merged</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--entity syntax--name syntax--function">mkdir</span> <span class="syntax--entity syntax--other syntax--attribute-name">-p</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OUT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">SAMPLE_MERGED</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--entity syntax--name syntax--function">mkdir</span> <span class="syntax--entity syntax--other syntax--attribute-name">-p</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OUT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">SAMPLE_MERGED</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/mapping</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--entity syntax--name syntax--function">mkdir</span> <span class="syntax--entity syntax--other syntax--attribute-name">-p</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OUT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">SAMPLE_MERGED</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/metrics</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--keyword syntax--control">if</span> [ <span class="syntax--entity syntax--other syntax--attribute-name">-e</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OUT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">SAMPLE_MERGED</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/mapping/<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">sample</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>.list ]</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--keyword syntax--control">then</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">                </span><span class="syntax--entity syntax--name syntax--function">rm</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OUT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">SAMPLE_MERGED</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/mapping/<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">sample</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>.list</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--keyword syntax--control">fi</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--keyword syntax--control">for</span> <span class="syntax--variable syntax--other syntax--member">i</span> <span class="syntax--keyword syntax--control">in</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">BAM_LIST</span>[@]:0:<span class="syntax--punctuation syntax--section syntax--embedded">${</span>#<span class="syntax--variable syntax--other syntax--member">BAM_LIST</span>[@]<span class="syntax--punctuation syntax--section syntax--embedded">}</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--keyword syntax--control">do</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">                </span><span class="syntax--entity syntax--name syntax--function"><span class="syntax--support syntax--function">echo</span></span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">i</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span> <span class="syntax--keyword syntax--operator">&gt;&gt;</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OUT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">SAMPLE_MERGED</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/mapping/<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">sample</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>.list</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--keyword syntax--control">done</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--entity syntax--name syntax--function">mkdir</span> <span class="syntax--entity syntax--other syntax--attribute-name">-p</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OUT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">SAMPLE_MERGED</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--entity syntax--name syntax--function">mkdir</span> <span class="syntax--entity syntax--other syntax--attribute-name">-p</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OUT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">SAMPLE_MERGED</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/mapping</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--entity syntax--name syntax--function">mkdir</span> <span class="syntax--entity syntax--other syntax--attribute-name">-p</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OUT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">SAMPLE_MERGED</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/metrics</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--entity syntax--name syntax--function">sbatch</span> <span class="syntax--entity syntax--other syntax--attribute-name">--cpus-per-task=1</span> <span class="syntax--entity syntax--other syntax--attribute-name">--mem-per-cpu=6G</span> \</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">                </span><span class="syntax--entity syntax--other syntax--attribute-name">-J</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">sample</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>_merging \</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">                </span><span class="syntax--entity syntax--other syntax--attribute-name">-o</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OUT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/logs/merging/<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">sample</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>_merging.o \</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">                </span><span class="syntax--entity syntax--other syntax--attribute-name">-e</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OUT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/logs/merging/<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">sample</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>_merging.e \</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">                </span><span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">SCRIPT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/mergingGenologin.sh <span class="syntax--entity syntax--other syntax--attribute-name">-b</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OUT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">SAMPLE_MERGED</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/mapping/<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">sample</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>.list <span class="syntax--entity syntax--other syntax--attribute-name">-o</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OUT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span> <span class="syntax--entity syntax--other syntax--attribute-name">-s</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">SAMPLE_MERGED</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span> <span class="syntax--entity syntax--other syntax--attribute-name">-t</span> 1</span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--keyword syntax--control">done</span> <span class="syntax--keyword syntax--operator">&lt;</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">SAMPLE_FILE</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#end of file</span></span></div></pre>
<h5 id="2432-sample_file-variable">2.4.3.2. SAMPLE_FILE variable</h5>
<p>List of samples to merge, in the form:</p>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function"><span class="syntax--punctuation syntax--section syntax--embedded">$</span> <span class="syntax--variable syntax--other syntax--member">head</span></span> <span class="syntax--entity syntax--other syntax--attribute-name">-3</span> SamplesToMerge.list</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">BER15</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">BER16</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">BER17</span></span></div></pre>
<h5 id="2432-startpoint-variable">2.4.3.2. STARTPOINT variable</h5>
<p>Path to the bams to be merged. As bam files are each in ~/SAMPLE_NAME/mapping directories (see <a href="#2310-output-files"> 2.3.10. Output FILES</a>), the paths are here in the form ~/DIRECTORY/*/*/mapping</p>
<h5 id="2433-other-variables">2.4.3.3. Other variables</h5>
<p>See <a href="#235-out-variable">2.3.5. OUT variable</a> and following</p>
<h4 id="2434-output-files">2.4.3.4 Output files</h4>
<p>For each sample, output files will be in a directory with the sample name (SAMPLE_NAME) in the form "SampleId_LibraryTag_merged" , each containing 1 directory, with the files:</p>
<ul>
<li>SAMPLE_NAME_sort.bam (before BQSR)</li>
</ul>
<p>For instance:</p>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function"><span class="syntax--punctuation syntax--section syntax--embedded">$</span> <span class="syntax--variable syntax--other syntax--member">ls</span></span> OUE8_CCGTCC_merged/mapping</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">OUE8_CCGTCC_merged_sort.bam</span>  OUE8_CCGTCC_merged_sort.bam.bai  OUE8.list</span></div></pre>
<p>These will have to be run through the BQSR and calling processes with the script Bootstrap_Call.bash</p>
<h4 id="244-bqsr-and-calling-on-the-merged-bams">2.4.4. BQSR and calling on the merged bams</h4>
<h5 id="2441-bootstrap_callbash-script">2.4.4.1. Bootstrap_Call.bash script</h5>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#!/bin/bash</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block"># Usage:    Bootstrap_Call.bash</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--variable syntax--other syntax--member">N</span>=1</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--variable syntax--other syntax--member">PLOIDY</span>=2</span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#------------------------------------------------------------------</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#EDIT FOR ACCESS TO SEQUENCE FILES</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--variable syntax--other syntax--member">SAMPLE_FILE</span>=/work/project/cytogen/Alain/seqapipopOnHAV3_1/Merging/forBootstrapCalling.list</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--variable syntax--other syntax--member">OUT</span>=/work/project/cytogen/Alain/seqapipopOnHAV3_1/Merging/MergedBams</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--variable syntax--other syntax--member">REF</span>=/home/gencel/vignal/save/Genomes/Abeille/HAv3_1_indexes/GCF_003254395.2_Amel_HAv3.1_genomic.fna</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--variable syntax--other syntax--member">CHROMOSOME</span>=/home/gencel/vignal/save/Genomes/Abeille/HAv3_1_indexes/HAv3_1_Chromosomes.list</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--variable syntax--other syntax--member">UNKNOWN</span>=/home/gencel/vignal/save/Genomes/Abeille/HAv3_1_indexes/HAv3_1_Unknown.list</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--variable syntax--other syntax--member">SCRIPT</span>=/genphyse/cytogen/seqapipop/ScriptsHAV3Mapping/mapping_calling</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#------------------------------------------------------------------</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">mkdir</span> <span class="syntax--entity syntax--other syntax--attribute-name">-p</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OUT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/logs/mapping    <span class="syntax--comment syntax--block"># /work/ktabetaoul/Project_seqapipop/RESULTATS/logs/mapping</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function"><span class="syntax--support syntax--function">cd</span></span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OUT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#EDIT FOR mapping.sh</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block"># This is the loop to send each sample to the cluster to be mapped</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--keyword syntax--control">while</span> <span class="syntax--entity syntax--name syntax--function">read</span> line</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--keyword syntax--control">do</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">    </span><span class="syntax--variable syntax--other syntax--member">ID</span>=`<span class="syntax--entity syntax--name syntax--function">basename</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">line</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>`    <span class="syntax--comment syntax--block"># give the name of the sample-population to map</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">    </span><span class="syntax--variable syntax--other syntax--member">IN</span>=`<span class="syntax--entity syntax--name syntax--function">dirname</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">line</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>`    <span class="syntax--comment syntax--block"># give the path where find the sample fastq</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">    </span><span class="syntax--entity syntax--name syntax--function">mkdir</span> <span class="syntax--entity syntax--other syntax--attribute-name">-p</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OUT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">ID</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">    </span><span class="syntax--entity syntax--name syntax--function">sbatch</span> <span class="syntax--entity syntax--other syntax--attribute-name">--cpus-per-task=1</span> <span class="syntax--entity syntax--other syntax--attribute-name">--mem-per-cpu=6G</span> \</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--entity syntax--other syntax--attribute-name">-J</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">ID</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>_mapping <span class="syntax--entity syntax--other syntax--attribute-name">-o</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OUT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/logs/<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">ID</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>_mapping.o <span class="syntax--entity syntax--other syntax--attribute-name">-e</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OUT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/logs/<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">ID</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>_mapping.e \</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">SCRIPT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/mappingAV_2019_Dec_RelanceBoots.sh <span class="syntax--entity syntax--other syntax--attribute-name">-s</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">ID</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span> <span class="syntax--entity syntax--other syntax--attribute-name">-i</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">IN</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span> <span class="syntax--entity syntax--other syntax--attribute-name">-o</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OUT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">ID</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span> <span class="syntax--entity syntax--other syntax--attribute-name">-p</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">PLOIDY</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span> <span class="syntax--entity syntax--other syntax--attribute-name">-n</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">N</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span> <span class="syntax--entity syntax--other syntax--attribute-name">-e</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OUT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/logs <span class="syntax--entity syntax--other syntax--attribute-name">-R</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">REF</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span> <span class="syntax--entity syntax--other syntax--attribute-name">-C</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">CHROMOSOME</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span> <span class="syntax--entity syntax--other syntax--attribute-name">-U</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">UNKNOWN</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--keyword syntax--control">done</span> <span class="syntax--keyword syntax--operator">&lt;</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">SAMPLE_FILE</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span></span></div><div class="line"></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block"># end of file</span></span></div></pre>
<h5 id="2442-the-sample_file-variable">2.4.4.2. The SAMPLE_FILE variable</h5>
<p>A list in the form:</p>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function"><span class="syntax--punctuation syntax--section syntax--embedded">$</span> <span class="syntax--variable syntax--other syntax--member">head</span></span> <span class="syntax--entity syntax--other syntax--attribute-name">-3</span> forBootstrapCalling.list</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">/work/project/cytogen/Alain/seqapipopOnHAV3_1/Merging/MergedBams/BER15_ATCACG_merged</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">/work/project/cytogen/Alain/seqapipopOnHAV3_1/Merging/MergedBams/BER16_CGATGT_merged</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">/work/project/cytogen/Alain/seqapipopOnHAV3_1/Merging/MergedBams/BER17_TTAGGC_merged</span></span></div></pre>
<h5 id="2442-other-variables">2.4.4.2. Other variables</h5>
<p>See <a href="#235-out-variable">2.3.5. OUT variable</a> and following</p>
<h4 id="2443-output-files">2.4.4.3. Output files</h4>
<p>For each sample, output files will be in a directory with the sample name (SAMPLE_NAME) in the form "SampleId_LibraryTag_merged", each containing 4 directories, with the bam and gvcf files:</p>
<ul>
<li>SAMPLE_NAME_sort.bam (before BQSR)</li>
<li>SAMPLE_NAME.bam (after BQSR)</li>
<li>SAMPLE_NAME.g.vcf.gz</li>
<li>SAMPLE_NAME_dedup.metrics</li>
</ul>
<p>For instance:</p>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function"><span class="syntax--punctuation syntax--section syntax--embedded">$</span> <span class="syntax--variable syntax--other syntax--member">ls</span></span> OUE8_CCGTCC_merged/*/</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">OUE8_CCGTCC_merged/bootstraping/:</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">OUE8_CCGTCC_merged.bam</span>  OUE8_CCGTCC_merged.bam.bai  OUE8_CCGTCC_merged_bam.list</span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">OUE8_CCGTCC_merged/calling/:</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">OUE8_CCGTCC_merged.g.vcf.gz</span>  OUE8_CCGTCC_merged.g.vcf.gz.tbi</span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">OUE8_CCGTCC_merged/mapping/:</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">OUE8_CCGTCC_merged_sort.bam</span>  OUE8_CCGTCC_merged_sort.bam.bai  OUE8.list</span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">OUE8_CCGTCC_merged/metrics/:</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">OUE8_CCGTCC_merged_dedup.metrics</span></span></div></pre>
<h4 id="2444-same-samples-under-different-names">2.4.4.4. Same samples under different names.</h4>
<p>JFM8 and JFM11 were used for testing a PCR-)free library preparation protocol and are sequenced under different names:</p>
<ul>
<li>JFM8_ACTTGA_L001</li>
<li>JFM8-PCRfree_TCGAAG_L001</li>
<li>JFM11_GGCTAC_L001</li>
<li>JFM11-PCRfree_TCGGCA_L001</li>
</ul>
<p>This causes problems in the calling step, as the sample names are different (merging works, but message at the calling step saying that there are two samples in the bam).</p>
<p>Two specific scripts had to be used: mergingJFM8.sh and mergingJFM11.sh, followed by Bootstrap_Call.bash</p>
<h3 id="25-controlling-that-all-gvcf-files-are-complete">2.5 controlling that all gvcf files are complete</h3>
<h4 id="251-script-controlvcfsbash">2.5.1 Script controlVcfs.bash</h4>
<ul>
<li>Counts the number of variants per chromosomes for each sample</li>
<li>All directories for all 870 samples are in ~/Haploid</li>
</ul>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#!/bin/bash</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#Usage: controlVcfs.bash</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">module</span> load <span class="syntax--entity syntax--other syntax--attribute-name">-f</span>  /home/gencel/vignal/save/000_ProgramModules/program_module</span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--keyword syntax--control">for</span> <span class="syntax--variable syntax--other syntax--member">i</span> <span class="syntax--keyword syntax--control">in</span> `<span class="syntax--entity syntax--name syntax--function">ls</span> /genphyse/cytogen/seqapipop/Data/Apis-mellifera/seqapipopOnHAV3_1/Haploid/*/calling/*.gz`</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--keyword syntax--control">do</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--variable syntax--other syntax--member">ID</span>=`<span class="syntax--entity syntax--name syntax--function">basename</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">i</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>`      <span class="syntax--comment syntax--block"># give the name of the sample-population to map</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--variable syntax--other syntax--member">IN</span>=`<span class="syntax--entity syntax--name syntax--function">dirname</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">i</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>`</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--entity syntax--name syntax--function">sbatch</span> <span class="syntax--entity syntax--other syntax--attribute-name">-J</span> test <span class="syntax--entity syntax--other syntax--attribute-name">--mem=1G</span> <span class="syntax--entity syntax--other syntax--attribute-name">--wrap=</span><span class="syntax--string">"module load -f /home/gencel/vignal/save/000_ProgramModules/program_module; \</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--string"><span class="leading-whitespace">        </span>bcftools query -f '%CHROM\n' <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">i</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span> | \</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--string"><span class="leading-whitespace">        </span>grep ^NC | uniq -c &gt; \</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--string"><span class="leading-whitespace">        </span>/work/project/cytogen/Alain/seqapipopOnHAV3_1/controlVcfs/<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">ID</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>.count"</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--keyword syntax--control">done</span></span></div></pre>
<h4 id="252-check-all-went-to-the-end">2.5.2. Check all went to the end</h4>
<ul>
<li>The last chromosome in the files is the mitochondrial DNA: NC_001566.1</li>
</ul>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function"><span class="syntax--punctuation syntax--section syntax--embedded">$</span> <span class="syntax--variable syntax--other syntax--member">grep</span></span> NC_001566.1 *.count <span class="syntax--keyword syntax--operator">|</span> <span class="syntax--entity syntax--name syntax--function">wc</span> <span class="syntax--entity syntax--other syntax--attribute-name">-l</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">870</span></span></div></pre>
<h2 id="3-making-the-large-vcf-file-with-all-870-samples">3. Making the large vcf file with all 870 samples</h2>
<p>In theory, the GATK CombineGVCFs accepts a list of samples, but I couldn't get it to work. Terefore, the bash script for combining all gvcf files will have all files listed in the form:</p>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">--variant</span> ~/Haploid/AOC16_GTTTCG_L002/calling/AOC16_GTTTCG_L002.g.vcf.gz \</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--other syntax--attribute-name">--variant</span> ~/Haploid/AOC17_CGTACG_L002/calling/AOC17_CGTACG_L002.g.vcf.gz \</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--other syntax--attribute-name">--variant</span> ~/Haploid/AOC18_GAGTGG_L002/calling/AOC18_GAGTGG_L002.g.vcf.gz \</span></div><div class="line"><span class="syntax--source syntax--shell">...</span></div></pre>
<h3 id="31-combine-gvcf-files">3.1 Combine gvcf files</h3>
<h3 id="311-generate-list-of-all-samples-for-the-combine-script">3.1.1. Generate list of all samples for the combine script</h3>
<ul>
<li>All directories for all 870 samples are in ~/Haploid</li>
</ul>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">ls</span> ~/Haploid/*/calling/* <span class="syntax--keyword syntax--operator">|</span> <span class="syntax--entity syntax--name syntax--function">grep</span> vcf.gz<span class="syntax--punctuation syntax--section syntax--embedded">$</span> <span class="syntax--keyword syntax--operator">|</span> <span class="syntax--entity syntax--name syntax--function">awk</span> <span class="syntax--string">'{print "    --variant",$1,"\\"}'</span> <span class="syntax--keyword syntax--operator">&gt;</span> temp.list</span></div></pre>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function"><span class="syntax--punctuation syntax--section syntax--embedded">$</span> <span class="syntax--variable syntax--other syntax--member">head</span></span> <span class="syntax--entity syntax--other syntax--attribute-name">-3</span> temp1.listed</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">--variant</span> ~/Haploid/AOC16_GTTTCG_L002/calling/AOC16_GTTTCG_L002.g.vcf.gz \</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--other syntax--attribute-name">--variant</span> ~/Haploid/AOC17_CGTACG_L002/calling/AOC17_CGTACG_L002.g.vcf.gz \</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--other syntax--attribute-name">--variant</span> ~/Haploid/AOC18_GAGTGG_L002/calling/AOC18_GAGTGG_L002.g.vcf.gz \</span></div></pre>
<h3 id="312-write-the-combining-scripts">3.1.2 Write the combining scripts</h3>
<blockquote>
<p>Done by combining Head (combineScriptHead) and Tail (combineScriptTail) files to temp.list</p>
</blockquote>
<ul>
<li>Head</li>
</ul>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function"><span class="syntax--punctuation syntax--section syntax--embedded">$</span> <span class="syntax--variable syntax--other syntax--member">more</span></span> combineScriptHead</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#!/bin/bash</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#combineGVCFsHAV3Called_slurm.bash</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">module</span> load <span class="syntax--entity syntax--other syntax--attribute-name">-f</span>  /home/gencel/vignal/save/000_ProgramModules/program_module</span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--keyword syntax--control">while</span> <span class="syntax--entity syntax--name syntax--function"><span class="syntax--support syntax--function">getopts</span></span> <span class="syntax--string">":c:"</span> opt; <span class="syntax--keyword syntax--control">do</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">  </span><span class="syntax--keyword syntax--control">case</span> <span class="syntax--punctuation syntax--section syntax--embedded">$</span><span class="syntax--variable syntax--other syntax--member">opt</span> <span class="syntax--keyword syntax--control">in</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">    </span>c) <span class="syntax--variable syntax--other syntax--member">CHR</span>=<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OPTARG</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>;;</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">  </span><span class="syntax--keyword syntax--control">esac</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--keyword syntax--control">done</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">gatk</span> <span class="syntax--entity syntax--other syntax--attribute-name">--java-options</span> <span class="syntax--string">"-Xmx80g"</span> CombineGVCFs  \</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--entity syntax--other syntax--attribute-name">-R</span> /home/gencel/vignal/save/Genomes/Abeille/HAv3_1_indexes/GCF_003254395.2_Amel_HAv3.1_genomic.fna \</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--entity syntax--other syntax--attribute-name">-L</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">CHR</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span> \</span></div></pre>
<ul>
<li>Tail</li>
</ul>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function"><span class="syntax--punctuation syntax--section syntax--embedded">$</span> <span class="syntax--variable syntax--other syntax--member">more</span></span> combineScriptTail</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--entity syntax--name syntax--function">-O</span> /work/project/cytogen/Alain/seqapipopOnHAV3_1/combineGVCFs/The870vcf/MetaGenotypes<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">CHR</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>.g.vcf.gz</span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function"><span class="syntax--support syntax--function">echo</span></span> <span class="syntax--string">"Finnished: "</span>`<span class="syntax--entity syntax--name syntax--function">date</span>`</span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#end of file</span></span></div></pre>
<ul>
<li>Combining together</li>
</ul>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">cat</span> combineScriptHead ../temp.list combineScriptTail <span class="syntax--keyword syntax--operator">&gt;</span> combineGVCFsHAV3_1_Called_slurm.bash</span></div></pre>
<h3 id="313-run-the-combine-script-with-combinegvcfshav3_1_lance_slurmbash">3.1.3 Run the combine script with combineGVCFsHAV3_1_Lance_slurm.bash</h3>
<ul>
<li>The script will work in parallel by combining the gvcf files for each chromosome separately</li>
</ul>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#!/bin/bash</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#combineGVCFsHAV3_1_Lance_slurm.bash</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--keyword syntax--control">for</span> <span class="syntax--variable syntax--other syntax--member">i</span> <span class="syntax--keyword syntax--control">in</span> `<span class="syntax--entity syntax--name syntax--function">cat</span> /home/gencel/vignal/save/Genomes/Abeille/HAv3_1_indexes/HAv3_1_Chromosomes.list <span class="syntax--keyword syntax--operator">|</span> <span class="syntax--entity syntax--name syntax--function">cut</span> <span class="syntax--entity syntax--other syntax--attribute-name">-f</span> 1`</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--keyword syntax--control">do</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">mkdir</span> <span class="syntax--entity syntax--other syntax--attribute-name">-p</span> ~/seqapipopOnHAV3_1/combineGVCFs/The870vcf/logs</span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">sbatch</span> <span class="syntax--entity syntax--other syntax--attribute-name">--cpus-per-task=1</span> <span class="syntax--entity syntax--other syntax--attribute-name">--mem-per-cpu=100G</span> \</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--entity syntax--other syntax--attribute-name">-J</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">i</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>_genotAll \</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--entity syntax--other syntax--attribute-name">-o</span> ~/seqapipopOnHAV3_1/combineGVCFs/The870vcf/logs/<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">i</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>_combine.o \</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--entity syntax--other syntax--attribute-name">-e</span> ~/seqapipopOnHAV3_1/combineGVCFs/The870vcf/logs/<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">i</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>_combine.e \</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span>/seqapipopOnHAV3_1/combineGVCFs/The870vcf/combineGVCFsHAV3_1_Called_slurm.bash <span class="syntax--entity syntax--other syntax--attribute-name">-c</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">i</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--keyword syntax--control">done</span></span></div></pre>
<h3 id="314-output">3.1.4. Output</h3>
<ul>
<li>One file per chromosome: 16 autosomes, plus mitochondrial DNA, plus Unknown, in the form:<ul>
<li>MetaGenotypesNC_001566.1.g.vcf.gz</li>
</ul>
</li>
</ul>
<h3 id="315-check-all-went-to-the-end">3.1.5. Check all went to the end</h3>
<ul>
<li>The idea is to printout the last positions for each chromosome's gvcf file.</li>
<li><h4 id="3151-example-for-mitochondrial-dna">3.1.5.1. Example for mitochondrial DNA</h4>
</li>
</ul>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">vignal@genologin2</span> ~/seqapipopOnHAV3_1/combineGVCFs/The870vcf <span class="syntax--punctuation syntax--section syntax--embedded">$</span> <span class="syntax--variable syntax--other syntax--member">bcftools</span> query <span class="syntax--entity syntax--other syntax--attribute-name">-f</span> <span class="syntax--string">'%CHROM\t%POS\n'</span> MetaGenotypesNC_001566.1.g.vcf.gz <span class="syntax--keyword syntax--operator">|</span> <span class="syntax--entity syntax--name syntax--function">tail</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_001566.1</span>     16331</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_001566.1</span>     16332</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_001566.1</span>     16333</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_001566.1</span>     16334</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_001566.1</span>     16335</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_001566.1</span>     16337</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_001566.1</span>     16338</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_001566.1</span>     16339</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_001566.1</span>     16342</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_001566.1</span>     16343</span></div></pre>
<ul>
<li>OK, mitochondrial chromosome is 16471 bp long</li>
</ul>
<h4 id="3152-for-all-chromosomes">3.1.5.2. For all chromosomes:</h4>
<ul>
<li>script combineCheck.bash</li>
</ul>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#!/bin/bash</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#combineCheck.bash</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">module</span> load <span class="syntax--entity syntax--other syntax--attribute-name">-f</span>  /home/gencel/vignal/save/000_ProgramModules/program_module</span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--keyword syntax--control">for</span> <span class="syntax--variable syntax--other syntax--member">i</span> <span class="syntax--keyword syntax--control">in</span> `<span class="syntax--entity syntax--name syntax--function">ls</span> ~/seqapipopOnHAV3_1/combineGVCFs/The870vcf/*.gz`</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--keyword syntax--control">do</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--variable syntax--other syntax--member">ID</span>=`<span class="syntax--entity syntax--name syntax--function">basename</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">i</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>`</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--variable syntax--other syntax--member">OUT</span>=<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">ID</span>/<span class="syntax--string syntax--regexp">gz</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>check</span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--entity syntax--name syntax--function"><span class="syntax--support syntax--function">echo</span></span>  <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">ID</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--entity syntax--name syntax--function"><span class="syntax--support syntax--function">echo</span></span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">ID</span>/<span class="syntax--string syntax--regexp">gz</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>check</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--entity syntax--name syntax--function"><span class="syntax--support syntax--function">echo</span></span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">i</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">sbatch</span> <span class="syntax--entity syntax--other syntax--attribute-name">--cpus-per-task=1</span> <span class="syntax--entity syntax--other syntax--attribute-name">--mem-per-cpu=1G</span> \</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--entity syntax--other syntax--attribute-name">--wrap=</span><span class="syntax--string">"module load -f /home/gencel/vignal/save/000_ProgramModules/program_module; \</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--string"><span class="leading-whitespace">        </span>bcftools query -f '%CHROM\t%POS\n' <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">i</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span> | tail  &gt; \</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--string"><span class="leading-whitespace">        </span>~/Alain/seqapipopOnHAV3_1/combineGVCFs/The870vcf/checkCombines/<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OUT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>"</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--keyword syntax--control">done</span></span></div></pre>
<ul>
<li>Outputs the last 10 positions for each chromosome in a file with names in the form: MetaGenotypesNC_001566.1.g.vcf.check</li>
<li>Positions of last variant for each chromosome:</li>
</ul>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function"><span class="syntax--punctuation syntax--section syntax--embedded">$</span> <span class="syntax--variable syntax--other syntax--member">for</span></span> i in `<span class="syntax--entity syntax--name syntax--function">ls</span> *.check`; <span class="syntax--entity syntax--name syntax--function">do</span> tail <span class="syntax--entity syntax--other syntax--attribute-name">-1</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">i</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>; <span class="syntax--entity syntax--name syntax--function">done</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_001566.1</span>     16343</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037638.1</span>     27752935</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037639.1</span>     16088822</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037640.1</span>     13618309</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037641.1</span>     13404430</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037642.1</span>     13896581</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037643.1</span>     17789025</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037644.1</span>     14198576</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037645.1</span>     12716507</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037646.1</span>     12354641</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037647.1</span>     12360026</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037648.1</span>     16352587</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037649.1</span>     11514127</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037650.1</span>     11279598</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037651.1</span>     10669571</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037652.1</span>     9533787</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037653.1</span>     7237125</span></div></pre>
<ul>
<li>Compare to chromosome lengths:</li>
</ul>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function"><span class="syntax--punctuation syntax--section syntax--embedded">$</span> <span class="syntax--variable syntax--other syntax--member">cat</span></span> /home/gencel/vignal/save/Genomes/Abeille/HAv3_1_indexes/HAv3_1_Chromosomes.list</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037638.1</span>     27754200</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037639.1</span>     16089512</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037640.1</span>     13619445</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037641.1</span>     13404451</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037642.1</span>     13896941</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037643.1</span>     17789102</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037644.1</span>     14198698</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037645.1</span>     12717210</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037646.1</span>     12354651</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037647.1</span>     12360052</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037648.1</span>     16352600</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037649.1</span>     11514234</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037650.1</span>     11279722</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037651.1</span>     10670842</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037652.1</span>     9534514</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037653.1</span>     7238532</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_001566.1</span>     16343</span></div></pre>
<h3 id="32-genotype-across-the-combined-gvcf">3.2. Genotype across the Combined GVCF</h3>
<h4 id="321-genotype">3.2.1. Genotype</h4>
<ul>
<li>Genotyping script genotypeGVCFsHAV3_1_Called_slurm.bash</li>
</ul>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#!/bin/bash</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#genotypeGVCFsHAV3_1_Called_slurm.bash</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">module</span> load <span class="syntax--entity syntax--other syntax--attribute-name">-f</span>  /home/gencel/vignal/save/000_ProgramModules/program_module</span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--keyword syntax--control">while</span> <span class="syntax--entity syntax--name syntax--function"><span class="syntax--support syntax--function">getopts</span></span> <span class="syntax--string">":c:"</span> opt; <span class="syntax--keyword syntax--control">do</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">  </span><span class="syntax--keyword syntax--control">case</span> <span class="syntax--punctuation syntax--section syntax--embedded">$</span><span class="syntax--variable syntax--other syntax--member">opt</span> <span class="syntax--keyword syntax--control">in</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">    </span>c) <span class="syntax--variable syntax--other syntax--member">CHR</span>=<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OPTARG</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>;;</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">  </span><span class="syntax--keyword syntax--control">esac</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--keyword syntax--control">done</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">gatk</span> <span class="syntax--entity syntax--other syntax--attribute-name">--java-options</span> <span class="syntax--string">"-Xmx80g"</span> GenotypeGVCFs  \</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--entity syntax--other syntax--attribute-name">-R</span> /home/gencel/vignal/save/Genomes/Abeille/HAv3_1_indexes/GCF_003254395.2_Amel_HAv3.1_genomic.fna \</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--entity syntax--other syntax--attribute-name">-V</span> /work/project/cytogen/Alain/seqapipopOnHAV3_1/combineGVCFs/The870vcf/MetaGenotypes<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">CHR</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>.g.vcf.gz \</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--entity syntax--other syntax--attribute-name">--use-new-qual-calculator</span> \</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">        </span><span class="syntax--entity syntax--other syntax--attribute-name">-O</span> /work/project/cytogen/Alain/seqapipopOnHAV3_1/combineGVCFs/The870vcf/MetaGenotypesCalled<span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">CHR</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>.vcf.gz</span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function"><span class="syntax--support syntax--function">echo</span></span> <span class="syntax--string">"Finnished: "</span>`<span class="syntax--entity syntax--name syntax--function">date</span>`</span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#end of file</span></span></div></pre>
<ul>
<li>Called by combineGVCFsHAV3_1_Lance_slurm.bash</li>
</ul>
<pre class="editor-colors lang-{bash,"><div class="line"><span class="syntax--text syntax--plain syntax--null-grammar">#!/bin/bash</span></div><div class="line"><span class="syntax--text syntax--plain syntax--null-grammar"></span></div><div class="line"><span class="syntax--text syntax--plain syntax--null-grammar">#genotypeGVCFsHAV3_1_Lance_slurm.bash</span></div><div class="line"><span class="syntax--text syntax--plain syntax--null-grammar"></span></div><div class="line"><span class="syntax--text syntax--plain syntax--null-grammar">for i in `cat /home/gencel/vignal/save/Genomes/Abeille/HAv3_1_indexes/HAv3_1_Chromosomes.list | cut -f 1`</span></div><div class="line"><span class="syntax--text syntax--plain syntax--null-grammar"></span></div><div class="line"><span class="syntax--text syntax--plain syntax--null-grammar">do</span></div><div class="line"><span class="syntax--text syntax--plain syntax--null-grammar">mkdir -p /work/project/cytogen/Alain/seqapipopOnHAV3_1/combineGVCFs/The870vcf/logsCalled</span></div><div class="line"><span class="syntax--text syntax--plain syntax--null-grammar"></span></div><div class="line"><span class="syntax--text syntax--plain syntax--null-grammar">sbatch --cpus-per-task=1 --mem-per-cpu=100G \</span></div><div class="line"><span class="syntax--text syntax--plain syntax--null-grammar"><span class="leading-whitespace">        </span>-J ${i}_genotAll \</span></div><div class="line"><span class="syntax--text syntax--plain syntax--null-grammar"><span class="leading-whitespace">        </span>-o /work/project/cytogen/Alain/seqapipopOnHAV3_1/combineGVCFs/The870vcf/logsCalled/${i}_genotype.o \</span></div><div class="line"><span class="syntax--text syntax--plain syntax--null-grammar"><span class="leading-whitespace">        </span>-e /work/project/cytogen/Alain/seqapipopOnHAV3_1/combineGVCFs/The870vcf/logsCalled/${i}_genotype.e \</span></div><div class="line"><span class="syntax--text syntax--plain syntax--null-grammar"><span class="leading-whitespace">        </span>/work/project/cytogen/Alain/seqapipopOnHAV3_1/combineGVCFs/The870vcf/genotypeGVCFsHAV3_1_Called_slurm.bash -c ${i}</span></div><div class="line"><span class="syntax--text syntax--plain syntax--null-grammar">done</span></div></pre>
<h4 id="322-check-output">3.2.2. Check output</h4>
<p><code style="font-family: Menlo, Consolas, &quot;DejaVu Sans Mono&quot;, monospace;">$ grep complete *_genotype.e</code></p>
<p>Should find a line for each chromosome</p>
<h3 id="33-concatenate-vcf-files">3.3 Concatenate vcf files</h3>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#!/bin/bash</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#concatenateVCFs.bash</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">module</span> load <span class="syntax--entity syntax--other syntax--attribute-name">-f</span>  /home/gencel/vignal/save/000_ProgramModules/program_module</span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">bcftools</span> concat <span class="syntax--entity syntax--other syntax--attribute-name">-f</span> vcf.list <span class="syntax--entity syntax--other syntax--attribute-name">-o</span> MetaGenotypesCalled870.vcf.gz <span class="syntax--entity syntax--other syntax--attribute-name">-O</span> z</span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">tabix</span> MetaGenotypesCalled870.vcf.gz</span></div></pre>
<h3 id="34-count-variants-per-chromosomes">3.4. Count variants per chromosomes</h3>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#!/bin/bash</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#statsVcfsRaw.bash</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">zcat</span> MetaGenotypesCalled870.vcf.gz  <span class="syntax--keyword syntax--operator">|</span> <span class="syntax--entity syntax--name syntax--function">grep</span> <span class="syntax--entity syntax--other syntax--attribute-name">-v</span> <span class="syntax--string">'#'</span> <span class="syntax--keyword syntax--operator">|</span> <span class="syntax--entity syntax--name syntax--function">cut</span> <span class="syntax--entity syntax--other syntax--attribute-name">-f</span> 1 <span class="syntax--keyword syntax--operator">|</span> <span class="syntax--entity syntax--name syntax--function">sort</span> <span class="syntax--keyword syntax--operator">|</span> <span class="syntax--entity syntax--name syntax--function">uniq</span> <span class="syntax--entity syntax--other syntax--attribute-name">-c</span> <span class="syntax--keyword syntax--operator">|</span> <span class="syntax--entity syntax--name syntax--function">awk</span> <span class="syntax--string">'BEGIN {OFS="\t";sum=0}{print $2, $1; sum += $1} END {print "Sum", sum}'</span> <span class="syntax--keyword syntax--operator">&gt;</span> countVcfSumRaw</span></div></pre>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function"><span class="syntax--punctuation syntax--section syntax--embedded">$</span> <span class="syntax--variable syntax--other syntax--member">more</span></span> countVcfSumRaw</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_001566.1</span>     1510</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037638.1</span>     1908476</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037639.1</span>     1057375</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037640.1</span>     982972</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037641.1</span>     889447</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037642.1</span>     953841</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037643.1</span>     1203559</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037644.1</span>     1032802</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037645.1</span>     898131</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037646.1</span>     808198</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037647.1</span>     759176</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037648.1</span>     976036</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037649.1</span>     822893</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037650.1</span>     808625</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037651.1</span>     706809</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037652.1</span>     627851</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037653.1</span>     552873</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">Sum</span>     14990574</span></div></pre>
<ul>
<li>Almost 15 million variants!</li>
</ul>
<h3 id="35-retain-only-snps">3.5 Retain only SNPs</h3>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#!/bin/bash</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#retainSNPs.bash</span></span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">module</span> load <span class="syntax--entity syntax--other syntax--attribute-name">-f</span>  /home/gencel/vignal/save/000_ProgramModules/program_module</span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--variable syntax--other syntax--member">OUT</span>=~/seqapipopOnHAV3_1/combineGVCFs/The870vcf</span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">gatk</span> <span class="syntax--entity syntax--other syntax--attribute-name">--java-options</span> <span class="syntax--string">"-Xmx64g"</span> SelectVariants  \</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">    </span><span class="syntax--entity syntax--other syntax--attribute-name">-R</span> /home/gencel/vignal/save/Genomes/Abeille/HAv3_1_indexes/GCF_003254395.2_Amel_HAv3.1_genomic.fna \</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">    </span><span class="syntax--entity syntax--other syntax--attribute-name">-V</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OUT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/MetaGenotypesCalled870.vcf.gz \</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">    </span><span class="syntax--entity syntax--other syntax--attribute-name">--select-type-to-include</span> SNP \</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="leading-whitespace">    </span><span class="syntax--entity syntax--other syntax--attribute-name">-O</span> <span class="syntax--punctuation syntax--section syntax--embedded">${</span><span class="syntax--variable syntax--other syntax--member">OUT</span><span class="syntax--punctuation syntax--section syntax--embedded">}</span>/MetaGenotypesCalled870_raw_snps.vcf.gz</span></div><div class="line"></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--comment syntax--block">#Count the SNPs</span></span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">zcat</span> MetaGenotypesCalled870_raw_snps.vcf.gz  <span class="syntax--keyword syntax--operator">|</span> <span class="syntax--entity syntax--name syntax--function">grep</span> <span class="syntax--entity syntax--other syntax--attribute-name">-v</span> <span class="syntax--string">'#'</span> <span class="syntax--keyword syntax--operator">|</span> <span class="syntax--entity syntax--name syntax--function">cut</span> <span class="syntax--entity syntax--other syntax--attribute-name">-f</span> 1 <span class="syntax--keyword syntax--operator">|</span> <span class="syntax--entity syntax--name syntax--function">sort</span> <span class="syntax--keyword syntax--operator">|</span> <span class="syntax--entity syntax--name syntax--function">uniq</span> <span class="syntax--entity syntax--other syntax--attribute-name">-c</span> <span class="syntax--keyword syntax--operator">|</span> <span class="syntax--entity syntax--name syntax--function">awk</span> <span class="syntax--string">'BEGIN {OFS="\t";sum=0}{print $2, $1; sum += $1} END {print "Sum", sum}'</span> <span class="syntax--keyword syntax--operator">&gt;</span> countVcfSumRawSNPs</span></div></pre>
<pre class="editor-colors lang-bash"><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function"><span class="syntax--punctuation syntax--section syntax--embedded">$</span> <span class="syntax--variable syntax--other syntax--member">more</span></span> countVcfSumRawSNPs</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_001566.1</span>     1057</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037638.1</span>     1351213</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037639.1</span>     755896</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037640.1</span>     685281</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037641.1</span>     631431</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037642.1</span>     676224</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037643.1</span>     854365</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037644.1</span>     725683</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037645.1</span>     626552</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037646.1</span>     572812</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037647.1</span>     545884</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037648.1</span>     689572</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037649.1</span>     583414</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037650.1</span>     580887</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037651.1</span>     496859</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037652.1</span>     441923</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">NC_037653.1</span>     382401</span></div><div class="line"><span class="syntax--source syntax--shell"><span class="syntax--entity syntax--name syntax--function">Sum</span>     10601454</span></div></pre>
<ul>
<li>More than 10 million SNPs!</li>
<li>The bee genome is 250 Gb long =&gt; one SNP per 25 bp!</li>
</ul>
<p>end file</p></body>
      </html>
